### ISSUE TITLE
python version memory leak

### ISSUE THREAD

**Issue by SigmaStar** [fireeye/capa#736]
### Description

repeat call to capa.main.main(argv=args) would cause memory leak, although gc.collect would collect most rubbish but there are still 10MB RSS memory remain unclean.

### Steps to Reproduce

just call capa.main many times

**Expected behavior:**

release all memory allocated to capa.main

**Actual behavior:**

RSS memory consuming increased a lot.

### Versions

capa 2.0.0

### Additional Information

Nope

**Comment by williballenthin** [comment]
thanks for reporting this @SigmaStar!

it will take a little bit of work to profile the memory usage and i think its a worthwhile effort. we'll update here with the results of our triage and potential fixes.

**Comment by williballenthin** [comment]
recently when running the capa tests locally they've been getting OOM killed. i wonder if this is related.

![image](https://user-images.githubusercontent.com/156560/130697418-146ff3ac-5d22-476e-be55-bb8f40c924e6.png)

**Comment by williballenthin** [comment]
i'm unable to reproduce the memory leak, at least as recorded by `tracemalloc`:

![image](https://user-images.githubusercontent.com/156560/130701176-0d73e6cf-0c4a-41b1-9961-481dd9ae686d.png)

this output is generated by the script here: https://github.com/fireeye/capa/blob/master/scripts/profile-memory.py
it runs `capa.main.main()` repeatedly and then shows the memory usage right at the end of the program. the output indicates that running `capa.main.main()` once and ten times results in approximately the same memory usage.

@SigmaStar can you provide any additional detail that would help me triage this potential memory leak?

**Comment by williballenthin** [comment]
I suppose the profile runs above show the memory usage from a python perspective, while @SigmaStar records memory via OS/RSS numbers. if the memory is fragmented and cannot be returned to the OS then i suppose RSS might continue to grow? the outlook here would probably not be good, so lets find a way to (dis)prove this theory.

**Comment by williballenthin** [comment]
re-running tweaked script that displayed RSS along the way. 100 iterations. will have results in a bit...

![image](https://user-images.githubusercontent.com/156560/130702486-cd73f7c2-5baf-40e1-b14f-5fcf90ca77e1.png)

**Comment by SigmaStar** [comment]
Yes, of course. I think maybe some rules or signatures lead to that memory leak. I test my code on both windows 10 1903 and ubuntu and on both platform it seems there is RSS memory increasing. We uses public rules (I think it can be cloned from github). Right now I use subprocess module in python instead of calling capa.main and force the operating system to recycle all the memory allocated. I notice that each time capa.main will use about 700 MB memory. So ten times of iteration will eventually take away at least 6 GB memory.



---Original---
From: "Willi ***@***.***&gt;
Date: Wed, Aug 25, 2021 07:08 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [fireeye/capa] python version memory leak (#736)




 
i'm unable to reproduce the memory leak, at least as recorded by tracemalloc:
 

 
this output is generated by the script here: https://github.com/fireeye/capa/blob/master/scripts/profile-memory.py
 it runs capa.main.main() repeatedly and then shows the memory usage right at the end of the program. the output indicates that running capa.main.main() once and ten times results in approximately the same memory usage.
 
@SigmaStar can you provide any additional detail that would help me triage this potential memory leak?
 
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by SigmaStar** [comment]
It's odd. On my laptop it's i7-10875H cpu at frequency 4.5GHz. I specified the rules directory and each iteration costs approximately 5 min! So I think it must be some rules that causes capa failed to recycle memory allocated.



---Original---
From: "Willi ***@***.***&gt;
Date: Wed, Aug 25, 2021 07:23 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [fireeye/capa] python version memory leak (#736)




 
re-running tweaked script that displayed RSS along the way. 100 iterations. will have results in a bit...
 

 
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by williballenthin** [comment]
If you have a few minutes, can you try to reproduce the behavior of the `profile-memory.py` script? 

Once this run of 100 is done, then I'll also try under Windows (though I don't expect the results to be substantially different).

**Comment by williballenthin** [comment]
You can also try to copy the memory profiling code into your harness and see if it highlights the lines causing most allocations. 

https://github.com/fireeye/capa/blob/33c3c7e106e945a8a633f2dec06eba936a1e9cc9/scripts/profile-memory.py#L69

**Comment by SigmaStar** [comment]
Yes, of course. But where can I download profile-memory.py?



---Original---
From: "Willi ***@***.***&gt;
Date: Wed, Aug 25, 2021 07:45 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [fireeye/capa] python version memory leak (#736)




 
If you have a few minutes, can you try to reproduce the behavior of the profile-memory.py script?
 
Once this run of 100 is done, then I'll also try under Windows (though I don't expect the results to be substantially different).
 
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by williballenthin** [comment]
Incidentally, we also run capa as a subprocess in our clusters. This makes it easier to kill long running jobs without affecting other runs. Also tracking memory, CPU, etc. 

That being said, its not an excuse for memory leaks, so I'd like to address this one if we can find it.

**Comment by williballenthin** [comment]
> Yes, of course. But where can I download profile-memory.py?

https://github.com/fireeye/capa/blob/master/scripts/profile-memory.py#L69

**Comment by SigmaStar** [comment]
Ok, I will do that later and report the memory profiling as soon as possible.



---Original---
From: "Willi ***@***.***&gt;
Date: Wed, Aug 25, 2021 07:46 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [fireeye/capa] python version memory leak (#736)




 
You can also try to copy the memory profiling code into your harness and see if it highlights the lines causing most allocations.
 
https://github.com/fireeye/capa/blob/33c3c7e106e945a8a633f2dec06eba936a1e9cc9/scripts/profile-memory.py#L69
 
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by williballenthin** [comment]
run of 100 iterations:

![image](https://user-images.githubusercontent.com/156560/130704327-88acdac7-1289-4512-b186-0cfd55c67521.png)

![image](https://user-images.githubusercontent.com/156560/130704346-ca4f290c-9ac3-4a56-8206-4a53d0bac620.png)

![image](https://user-images.githubusercontent.com/156560/130704359-aaa3ec99-59d4-4b55-aeeb-4029fe022191.png)

![image](https://user-images.githubusercontent.com/156560/130704428-ea463c49-d64b-4449-93f5-6284d2ef77aa.png)

while there's some variation, its not clear if its an upward trend of 10MB over 100 runs or just noise. definitely not on the order of 10MB/run.

**Comment by SigmaStar** [comment]
![image](https://user-images.githubusercontent.com/22829962/130712323-95e327e6-29cd-455d-a0a1-8ceba1d905fa.png)
![image](https://user-images.githubusercontent.com/22829962/130712359-b8736530-4322-462a-8cf6-b62571b5a2c0.png)
This is the output on my  laptop (Windows 10 1903 with Miniconda python 3.8.5) and you can see the RSS increases for each run.

**Comment by williballenthin** [comment]
ahhh, this is very interesting and useful. the top 10 lines make it very clear that part of vivisect (specifically vtrace) is keeping a lot of memory around. i'll dig into what its doing and if there's a way we can:

  1. fix this upstream in viv, and
  2. fix this locally, as well, maybe by manually clearing out some fields

thanks for collecting and sharing these results!

**Comment by SigmaStar** [comment]
By the way I changed the script as :
![image](https://user-images.githubusercontent.com/22829962/130713704-6efe9641-7bec-4a7c-b8a8-ddb36e48c90f.png)
The rules are cloned from https://github.com/fireeye/capa-rules

**Comment by williballenthin** [comment]
yup, that makes sense and looks fine. i think the results are reliable and it was due to linux vs windows that i didn’t see the same thing. i’ll reproduce first thing tomorrow.

**Comment by williballenthin** [comment]
we are not able to reproduce this issue, so we won't be able to fix it. if we become able to reproduce the issue, then this will be a high priority item to address.
